lamagraph-compiler/src/LamaML/Syntax.hs

```diff
@@ -0,0 +1,290 @@
+{- |
+
+LamaML syntax
+-}
+module LamaML.Syntax (
+    -- * Grammar rules
+    -- $rules
+
+    -- * Lexical conventions
+
+    -- ** Blanks
+    -- $lexing_blanks
+
+    -- ** Comments
+    -- $lexing_comments
+
+    -- ** Identifiers
+    -- $lexing_idents
+
+    -- ** Integer literals
+    -- $lexing_int_lits
+
+    -- ** Character literals
+    -- $lexing_char_lits
+
+    -- ** String literals
+    -- $lexing_string_lits
+
+    -- ** Operators
+    -- $lexing_operators
+
+    -- ** Keywords
+    -- $lexing_keywords
+
+    -- * Names
+    -- $names
+
+    -- * Type expressions
+    -- $types
+
+    -- * Constants
+    -- $constants
+
+    -- * Patterns
+    -- $patterns
+
+    -- * Expressions
+    -- $expressions
+
+    -- * Type definitions
+    -- $typedefs
+
+    -- * Declarations
+    -- $decls
+) where
+
+{- $rules
+Grammar rules are written in monospace font.
+Italicized words represent @/nonterminals/@.
+@Terminals@ are written in upright font.
+@( )@ mean grouping, @{ }@ mean zero or more repetitions, @[ ]@ mean optional.
+In case of clash between grammar and terminal meaning, terminal symbol will be typeset as @__bold monospace__@.
+-}
+
+{- $lexing_blanks
+Spaces, tabs, @\\n@ and @\\r@ are ignored and are used to separate tokens.
+-}
+
+{- $lexing_comments
+Comments must be enclosed in @(*@ and @*)@.
+Nesting is allowed and must be handled.
+-}
+
+{- $lexing_idents
+@
+/ident/ ::= ( /letter/ | _ ) { /letter/ | 0...9 | _ | ' }
+
+/capitalized-ident/ ::= ( A...Z ) { /letter/ | 0...9 | _ | ' }
+
+/lowercase-ident/ ::= ( a...z | _ ) { /letter/ | 0...9 | _ | ' }
+
+/letter/ ::= A...Z | a...z
+@
+-}
+
+{- $lexing_int_lits
+@
+/integer-literal/ ::= \[-] ( 0...9 ) { 0...9 | _ }
+@
+
+/Do we need hex\/octal\/binary numbers and\/or size suffixes?/
+
+/Also no floats currently./
+-}
+
+{- $lexing_char_lits
+@
+/char-literal/ ::= ' /regular-char/ '
+               | ' /escape-sequence/ '
+
+/escape-sequence/ ::= \\ ( __|__ | " | ' | n )
+@
+
+/Do we need more escape sequences?/
+
+/We must define regular-char? Any ASCII one?/
+-}
+
+{- $lexing_string_lits
+@
+/string-literal/ ::= " { /string-character/ } "
+
+/string-character/ ::= /regular-string-character/
+                   | /escape-sequence/
+@
+
+/We must define regular-string-character? Any ASCII one?/
+
+/Do we need multiline strings?/
+-}
+
+{- $lexing_operators
+
+@
+/infix-symbol/ ::= ( = | \< | \> | @ | ^ | __|__ | & | + | - | * | / | $ | % ) { /operator-char/ }
+
+/prefix-symbol/ ::= ! { /operator-char/ }
+                | ( ? | ~ ) { /operator-char/ }+
+
+/operator-char/ ::= ! | $ | % | & | * | + | - | . | / | : | \< | = | \> | ? | @ | ^ | __|__ | ~
+@
+
+Copypasted from https://askra.de/software/ocaml-doc/4.02/lex.html#sec71, probably too complicated.
+-}
+
+{- $lexing_keywords
+Keyword table:
+
+@
+and asr begin else end false fun function
```

Comment:
We really want to support both ```fun``` and ```function```?

What is ```asr```? Do we really need it?

Is it really good idea use keywords instead of operators for logical operations?

Reply:
> We really want to support both `fun` and `function`?

Correct me if I am wrong, but aren't `fun` and `function` identical in the following sense:
```ocaml
let f = fun x y z -> expr
let f = function x -> function y -> function z -> expr
```
and we'll need this step of translation from `fun` to `function` in our AST? 

> What is `asr`? Do we really need it?

Arithmetic shift right. Maybe we don't :)

> Is it really good idea use keywords instead of operators for logical operations?

Maybe I read too much ASM, but I find mnemonically written logical operations much easier to read and memorize. I'd like to stick with this style.

Reply:
> > We really want to support both `fun` and `function`?
> 
> Correct me if I am wrong, but aren't `fun` and `function` identical in the following sense:
> 
> ```ocaml
> let f = fun x y z -> expr
> let f = function x -> function y -> function z -> expr
> ```
> 
> and we'll need this step of translation from `fun` to `function` in our IR?
> 
```function``` allows one to simplify pattern matching in lambdas: https://ocaml.org/manual/5.2/expr.html#sss:expr-function-definition  

So, it is just a syntactic sugar. I think it is not a big problem to support it, but I'm not sure that is should be done at the first stage.

> > What is `asr`? Do we really need it?
> 
> Arithmetic shift right. Maybe we don't :)
> 
> > Is it really good idea use keywords instead of operators for logical operations?
> 
> Maybe I read too much ASM, but I find mnemonically written logical operations much easier to read and memorize. I'd like to stick with this style.

Well. Let's do so.